modules/services/davmail.nix

```
@@ -0,0 +1,142 @@
+{ config, lib, pkgs, ... }:
+let
+
+  inherit (lib)
+    attrNames boolToString concatMap concatStringsSep isAttrs isBool
+    mapAttrsRecursive mkDefault mkEnableOption mkIf mkOption optionalAttrs
+    types;
+
+  cfg = config.services.davmail;
+
+  settingsType = with types;
+    oneOf [ (attrsOf settingsType) str int bool ] // {
+      description =
+        "DavMail settings type (str, int, bool or attribute set thereof)";
+    };
+
+  toStr = val: if isBool val then boolToString val else toString val;
+
+  linesForAttrs = attrs:
+    concatMap (name:
+      let value = attrs.${name};
+      in if isAttrs value then
+        map (line: name + "." + line) (linesForAttrs value)
+      else
+        [ "${name}=${toStr value}" ]) (attrNames attrs);
+
+  settingsFile = pkgs.writeText "davmail.properties"
+    (concatStringsSep "\n" (linesForAttrs cfg.settings));
```

I feel like this could all be simplified for generating the config. 

```suggestion
  settingsFile = pkgs.writeText "davmail.properties"
    (lib.generators.toKeyValue { indent = "  "; } cfg.settings));
```

---

I tried this, but it fails since `lib.generators.toKeyValue` does not handle nested attrs.

---

https://github.com/nix-community/home-manager/pull/6674#discussion_r2008924713

---

I see where you're getting at. Actually, there also is `pkgs.formats.javaProperties` which seems to be even better in this case. And interestingly, a [comment in the implementation of it](https://github.com/NixOS/nixpkgs/blob/f9ec383a105dfa6ea11dc7527563aa6e929503bf/pkgs/pkgs-lib/formats/java-properties/default.nix#L23-L45) discourages the use of nested attribute sets.

Hence, I refactored the module to use `pkgs.formats.javaProperties`.